src/control/cmd/agent/mgmt_rpc.go

```diff
@@ -63,14 +98,82 @@ func (mod *mgmtModule) ID() int32 {
 	return drpc.ModuleMgmt
 }
 
-func (mod *mgmtModule) handleGetAttachInfo(reqb []byte) ([]byte, error) {
+// loadBalance is a simple round-robin load balancing scheme
+// to assign network interface adapters to clients
+// on the same NUMA node that have multiple adapters
+// to choose from.
+func (mod *mgmtModule) loadBalance(numaNode uint) {
+	numDevs := len(mod.resmgmtpb[numaNode])
+	newDevIdx := (mod.devIdx[numaNode] + 1) % numDevs
+	mod.devIdx[numaNode] = newDevIdx
+}
+
+// handleGetAttachInfo invokes the GetAttachInfo dRPC.  The agent determines the
+// NUMA node for the client process based on its PID.  Then based on the
+// server's provider, chooses a matching network interface and domain from the
+// client machine that has the same NUMA affinity.  It is considered an error if
+// the client application is bound to a NUMA node that does not have a network
+// device / provider combination with the same NUMA affinity.
+//
+// The client machine may have more than one matching network interface per
+// NUMA node.  In order to load balance the client application's use of multiple
+// network interfaces on a given NUMA node, a round robin resource allocation
+// scheme is used to choose the next device for that node.  See "loadBalance()"
+// for details.
+//
+// The agent caches the local device data and all possible responses the first
+// time this dRPC is invoked. Subsequent calls receive the cached data.
+// The use of cached data may be disabled by exporting
+// "DAOS_AGENT_DISABLE_CACHE=true" in the environment running the daos_agent.
+func (mod *mgmtModule) handleGetAttachInfo(reqb []byte, pid int32) ([]byte, error) {
+	mod.mutex.Lock()
+	defer mod.mutex.Unlock()
+
 	if os.Getenv("DAOS_AGENT_DISABLE_CACHE") == "true" {
 		mod.cachedAttachInfo = false
 		mod.log.Debugf("GetAttachInfo agent caching has been disabled\n")
 	}
 
+	fi, err := os.Stat(fmt.Sprintf("/proc/%d", pid))
+	if err != nil {
+		return nil, errors.Wrapf(err, "client process %d not found", pid)
+	}
+
 	if mod.cachedAttachInfo {
-		return mod.resmgmtpb, nil
+		clientData := mod.clientData[pid]
+		// If it's a client with new PID, or a different client with a cached PID,
+		// update the cache with the new data.
+		if clientData.timeStamp != fi.ModTime().String() {
```

Comment:
The operation for looking up the NUMA node is slow.  I wanted to avoid having to do it more often than necessary.  By caching the response for a PID, it avoids the lookup, provided the data in the cache is still valid.

Reply:
But my point is that I don't see how the per-pid cached data will ever be reused. The DAOS client library linked to the process will request the attachinfo at startup, and then never again, right?

Reply:
Oh, I found that in the daos_test application, the same client makes numerous GetAttachInfo requests throughout its lifetime.  If it made it only once, I agree that the per-pid cached data won't get reused.  If daos_test is aberrant behavior, then sure, ignoring the per-pid data makes sense to simplify things.

Reply:
Just a quick followup - so I treated daos_test as an example of a normal client application.  If that's way off the charts crazy as far as client applications go, then I just got unlucky with these observations by applying it to the general case.

Reply:
If we decide not to cache the data based on PID, the load balancing is perturbed a bit.  Currently, a GetAttachInfo from the same PID does not cause the device index for that NUMA node to change.  The client gets the exact same information the Nth time as it did the first time.  If we don't cache the PID and therefore have nothing to compare the PID timestamp with, then the load balancer will be used to increment that NUMA node's device index each time a request comes in.  There is potential for the load balancing to be made unfair in this case.  For example, if there are 2 devices on NUMA 0, the first client on NUMA 0 to make the request gets device index 0.  If that same client makes a second request, the agent sees this is a new request, determines the client is NUMA 0 based on PID, and offers the next device on the list, which is device 1.  But, the client didn't relinquish device 0 because it's actually the same process as before.  It's not going to use the interface data this time, because it has already initialized CaRT by this time.  Then a second process on NUMA 0 makes the request while the first client is alive and running.  It should get device 1 for fairness, but the load balancer already gave device 1, and now hands out device 0.  Now there are two clients using device 0.  It's not a catastrophic problem.  We could choose to accept that we will ignore PID input into the cache selection and get the slightly wrong load balancing in favor of some slightly reduced code complexity.

Reply:
Hmm. I don't know offhand. Seems strange that it would request it more than once, though. I doubt that daos_test is itself requesting the attachinfo more than once, as it's just linking against the client API. I wonder if there's a bug that's causing the same process to request it more than once. @johannlombardi, @liw, any thoughts on this?

Reply:
It's an artifact of the pool create/connect (I forget which one).  It gets called a lot in daos_test.

Reply:
The DAOS API does not have the concept of a system handle. libdaos internally creates a dc_mgmt_sys object on demand when connecting to a pool. After disconnecting from all pools within this system, libdaos does not have a cache for the dc_mgmt_sys object, nor can there be a system handle referencing it. So the dc_mgmt_sys object is freed. Next time an operation needs to access this system, libdaos will inovke GetAttachInfo to create a new dc_mgmt_sys object.

Reply:
I feel this aspect of daos_test does not represent typical application uses.

Reply:
OK, I think it makes sense that the attachinfo request would be made when connecting to a pool, as it's possible that the management service leader may have changed in the time between process startup and pool (re-)connection, right?

For most production DAOS clients, I suspect that they will be connecting to only one pool for the duration of the process lifetime, and so it seems that creating a cache per client pid is an unnecessary optimization.

Reply:
Ok, I'll go ahead and rework it so that it no longer caches the response based on PID.  Instead, it will be cached purely by NUMA node.

Reply:
@mjmac, I think system attach info changes should be infrequent. libdaos doesn't cache it just because of simplicity. Since we know early on that we'll cache attach infos of systems in daos_agent, I figure daos_agent may be a better place to handle caching and potential infrequent invalidation. If necessary, libdaos could cache dc_mgmt_sys objects. Hope that clarifies the rational a bit.

Reply:
PID based caching has been removed and reworked.